src/busctl/busctl.c

```diff
@@ -1043,6 +1044,21 @@ static int introspect(int argc, char **argv, void *userdata) {
                         if (r < 0)
                                 return bus_log_parse_error(r);
 
+                        r = sd_bus_message_peek_type(reply, &type, &contents);
+                        if (r <= 0)
+                                return bus_log_parse_error(r == 0 ? EINVAL : r);
+
+                        if (type == SD_BUS_TYPE_STRUCT_BEGIN)
+                                signature = strjoin(CHAR_TO_STR(SD_BUS_TYPE_STRUCT_BEGIN), contents, CHAR_TO_STR(SD_BUS_TYPE_STRUCT_END));
+                        else if (type == SD_BUS_TYPE_DICT_ENTRY_BEGIN)
+                                signature = strjoin(CHAR_TO_STR(SD_BUS_TYPE_DICT_ENTRY_BEGIN), contents, CHAR_TO_STR(SD_BUS_TYPE_DICT_ENTRY_END));
+                        else if (contents)
+                                signature = strjoin(CHAR_TO_STR(type), contents);
+                        else
+                                signature = strdup(CHAR_TO_STR(type));
+                        if (!signature)
+                                return log_oom();
```

Comment:
that's quite baroque.

much simpler: peek the type before you enter the `v` container, above.

i.e. don't combine what was split for you, just use the combined type ebfore

Reply:
→ #26076